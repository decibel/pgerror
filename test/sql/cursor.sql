\set ECHO none

\i test/pgxntool/setup.sql

CREATE TEMP VIEW named_cursors AS SELECT * FROM pg_cursors WHERE name NOT LIKE '<unnamed portal %>';

CREATE TEMP TABLE test_table(
  s serial
  , i int
  , t text
);
INSERT INTO test_table(i,t) SELECT ser, ser::text FROM generate_series(10,1,-1) ser;

CREATE TEMP TABLE good(
  seq serial
  , code text
  , description text
);
INSERT INTO good(code, description) VALUES
  ('SELECT * FROM test_table ORDER BY i', 'ORDER BY i')
  , ('SELECT * FROM test_table ORDER BY s', 'ORDER BY s')
;

CREATE TEMP TABLE bad(
  code text
  , description text
  , sqlstate          text
  , message           text
  , hint              text
  , detail            text
  , context           text
  , schema_name       text
  , table_name        text
  , column_name       text
  , constraint_name   text
  , type_name         text
);
INSERT INTO bad VALUES
  ('bogus', 'syntax error'
    , '42601'
    , 'syntax error at or near "bogus"'
    , ''
    , ''
    , 'PL/pgSQL function try_cursor(text,text) line 6 at OPEN'
    , ''
    , ''
    , ''
    , ''
    , ''
  )
  -- TODO: Replace this with a function that raises specific items in the fields
  , ('SELECT * FROM pg_temp.non_existent_table', 'non existent table'
    , '42P01'
    , 'relation "pg_temp.non_existent_table" does not exist'
    , ''
    , ''
    , 'PL/pgSQL function try_cursor(text,text) line 6 at OPEN'
    , ''
    , ''
    , ''
    , ''
    , ''
  )
;

SELECT plan(
  0
  + 1 -- Verify no named cursors
  -- Unnamed bad
  + 11 * (SELECT count(*)::int FROM bad)
  -- Unnamed good
  + 2 * (SELECT count(*)::int FROM good)
  -- Named bad
  + 1 + 11 * (SELECT count(*)::int FROM bad)
  -- Named good
  + 1 + 2 * (SELECT count(*)::int FROM good)
);

SELECT is_empty(
  $$SELECT * FROM named_cursors$$
  , 'Verify there are no named cursors'
);


-- TODO: test other return values
SELECT ok(
      result IS NULL
      , description || ' result is NULL'
    )
    || E'\n' || is((error).sqlstate, bad.sqlstate, description || ' check sqlstate')
    || E'\n' || is((error).message, bad.message, description || ' check message')
    || E'\n' || is((error).hint, bad.hint, description || ' check hint')
    || E'\n' || is((error).detail, bad.detail, description || ' check detail')
    || E'\n' || is((error).context, bad.context, description || ' check context')
    || E'\n' || is((error).schema_name, bad.schema_name, description || ' check schema_name')
    || E'\n' || is((error).table_name, bad.table_name, description || ' check table_name')
    || E'\n' || is((error).column_name, bad.column_name, description || ' check column_name')
    || E'\n' || is((error).constraint_name, bad.constraint_name, description || ' check constraint_name')
    || E'\n' || is((error).type_name, bad.type_name, description || ' check type_name')
  FROM bad, try_cursor(code)
;

/*
 * Unnamed good
 */
SELECT results_eq(
      result
      , code
      , description || ' check results'
    )
    || E'\n' || is(
      error
      , NULL
      , description || ' error is null'
    )
  FROM good, try_cursor(code)
  ORDER BY seq
;

/*
 * Named bad
 */
SELECT ok(
      result IS NOT NULL
      , 'cursor named ' || description || ' result is NOT NULL'
    )
    || E'\n' || is((error).sqlstate, bad.sqlstate, 'cursor named ' || description || ' check sqlstate')
    || E'\n' || is((error).message, bad.message, 'cursor named ' || description || ' check message')
    || E'\n' || is((error).hint, bad.hint, 'cursor named ' || description || ' check hint')
    || E'\n' || is((error).detail, bad.detail, 'cursor named ' || description || ' check detail')
    || E'\n' || is((error).context, bad.context, 'cursor named ' || description || ' check context')
    || E'\n' || is((error).schema_name, bad.schema_name, 'cursor named ' || description || ' check schema_name')
    || E'\n' || is((error).table_name, bad.table_name, 'cursor named ' || description || ' check table_name')
    || E'\n' || is((error).column_name, bad.column_name, 'cursor named ' || description || ' check column_name')
    || E'\n' || is((error).constraint_name, bad.constraint_name, 'cursor named ' || description || ' check constraint_name')
    || E'\n' || is((error).type_name, bad.type_name, 'cursor named ' || description || ' check type_name')
  FROM bad, try_cursor(code, description)
;

SELECT is_empty(
  $$SELECT * FROM named_cursors$$
  , 'Verify there are no named cursors'
);

/*
 * Named good
 */
SELECT is(
        (try_cursor(code, description)).error
        , NULL
        , 'Good named cursor returns NULL error result'
      )
  FROM good
  ORDER BY seq
;

SELECT set_eq(
  $$SELECT name, statement FROM named_cursors$$
  , $$SELECT description, code FROM good$$
  , 'Verify named cursors exist'
);

SELECT results_eq(
      description::refcursor -- Now the name of a named cursor
      , code
      , 'named cursor ' || description || ' check results'
    )
  FROM good
  ORDER BY seq
;

\i test/pgxntool/finish.sql

-- vi: expandtab ts=2 sw=2
