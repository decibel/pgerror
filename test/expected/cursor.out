\set ECHO none
1..55
ok 1 - Verify there are no named cursors
ok 2 - syntax error result is NULL
ok 3 - syntax error check sqlstate
ok 4 - syntax error check message
ok 5 - syntax error check hint
ok 6 - syntax error check detail
ok 7 - syntax error check context
ok 8 - syntax error check schema_name
ok 9 - syntax error check table_name
ok 10 - syntax error check column_name
ok 11 - syntax error check constraint_name
ok 12 - syntax error check type_name
ok 13 - non existent table result is NULL
ok 14 - non existent table check sqlstate
ok 15 - non existent table check message
ok 16 - non existent table check hint
ok 17 - non existent table check detail
ok 18 - non existent table check context
ok 19 - non existent table check schema_name
ok 20 - non existent table check table_name
ok 21 - non existent table check column_name
ok 22 - non existent table check constraint_name
ok 23 - non existent table check type_name
ok 24 - ORDER BY i check results
ok 25 - ORDER BY i error is null
ok 26 - ORDER BY s check results
ok 27 - ORDER BY s error is null
ok 28 - cursor named syntax error result is NOT NULL
ok 29 - cursor named syntax error check sqlstate
ok 30 - cursor named syntax error check message
ok 31 - cursor named syntax error check hint
ok 32 - cursor named syntax error check detail
ok 33 - cursor named syntax error check context
ok 34 - cursor named syntax error check schema_name
ok 35 - cursor named syntax error check table_name
ok 36 - cursor named syntax error check column_name
ok 37 - cursor named syntax error check constraint_name
ok 38 - cursor named syntax error check type_name
ok 39 - cursor named non existent table result is NOT NULL
ok 40 - cursor named non existent table check sqlstate
ok 41 - cursor named non existent table check message
ok 42 - cursor named non existent table check hint
ok 43 - cursor named non existent table check detail
ok 44 - cursor named non existent table check context
ok 45 - cursor named non existent table check schema_name
ok 46 - cursor named non existent table check table_name
ok 47 - cursor named non existent table check column_name
ok 48 - cursor named non existent table check constraint_name
ok 49 - cursor named non existent table check type_name
ok 50 - Verify there are no named cursors
ok 51 - Good named cursor returns NULL error result
ok 52 - Good named cursor returns NULL error result
ok 53 - Verify named cursors exist
ok 54 - named cursor ORDER BY i check results
ok 55 - named cursor ORDER BY s check results
# TRANSACTION INTENTIONALLY LEFT OPEN!
